File: bench/mandelbrot/mandelbrot.go
package main

import (
	"fmt"
	"time"
)

const (
	w        = 256
	h        = 256
	maxIter  = 256
	nRuns    = 21
	nWarmup  = 5
	x0       = -2.0
	x1       = 0.5
	y0       = -1.25
	y1       = 1.25
)

func mix(h, x uint32) uint32 { return (h ^ x) * 0x01000193 }

func medianUs(samples []float64) int {
	for i := 1; i < len(samples); i++ {
		v := samples[i]
		j := i - 1
		for j >= 0 && samples[j] > v {
			samples[j+1] = samples[j]
			j--
		}
		samples[j+1] = v
	}
	return int(samples[(len(samples)-1)>>1] * 1000)
}

func checksumU32(xs []uint32) uint32 {
	h := uint32(0x811c9dc5)
	for i := 0; i < len(xs); i += 128 {
		h = mix(h, xs[i])
	}
	return h
}

func render(out []uint32) {
	dx := (x1 - x0) / float64(w)
	dy := (y1 - y0) / float64(h)
	for py := 0; py < h; py++ {
		cy := y0 + float64(py)*dy
		for px := 0; px < w; px++ {
			cx := x0 + float64(px)*dx
			var zx, zy float64
			i := 0
			for i < maxIter {
				x2 := zx * zx
				y2 := zy * zy
				if x2+y2 > 4.0 {
					break
				}
				zy = 2*zx*zy + cy
				zx = x2 - y2 + cx
				i++
			}
			out[py*w+px] = uint32(i)
		}
	}
}

func main() {
	out := make([]uint32, w*h)

	for i := 0; i < nWarmup; i++ {
		render(out)
	}
	samples := make([]float64, nRuns)
	for i := 0; i < nRuns; i++ {
		t0 := time.Now()
		render(out)
		samples[i] = float64(time.Since(t0).Nanoseconds()) / 1e6
	}
	fmt.Printf("median_us=%d checksum=%d samples=%d stages=%d runs=%d\n",
		medianUs(samples), checksumU32(out), w*h, maxIter, nRuns)
}

File: bench/mandelbrot/mandelbrot.rs
use std::time::Instant;

const W: usize = 256;
const H: usize = 256;
const MAX_ITER: u32 = 256;
const N_RUNS: usize = 21;
const N_WARMUP: usize = 5;

const X0: f64 = -2.0;
const X1: f64 = 0.5;
const Y0: f64 = -1.25;
const Y1: f64 = 1.25;

fn mix(h: u32, x: u32) -> u32 {
    (h ^ x).wrapping_mul(0x0100_0193)
}

fn median_us(samples: &mut [f64]) -> u64 {
    for i in 1..samples.len() {
        let v = samples[i];
        let mut j = i;
        while j > 0 && samples[j - 1] > v {
            samples[j] = samples[j - 1];
            j -= 1;
        }
        samples[j] = v;
    }
    (samples[(samples.len() - 1) >> 1] * 1000.0) as u64
}

fn checksum_u32(xs: &[u32]) -> u32 {
    let mut h = 0x811c_9dc5u32;
    let mut i = 0;
    while i < xs.len() { h = mix(h, xs[i]); i += 128; }
    h
}

fn render(out: &mut [u32]) {
    let dx = (X1 - X0) / W as f64;
    let dy = (Y1 - Y0) / H as f64;
    for py in 0..H {
        let cy = Y0 + py as f64 * dy;
        for px in 0..W {
            let cx = X0 + px as f64 * dx;
            let mut zx = 0.0f64;
            let mut zy = 0.0f64;
            let mut i = 0u32;
            while i < MAX_ITER {
                let x2 = zx * zx;
                let y2 = zy * zy;
                if x2 + y2 > 4.0 { break; }
                zy = 2.0 * zx * zy + cy;
                zx = x2 - y2 + cx;
                i += 1;
            }
            out[py * W + px] = i;
        }
    }
}

fn main() {
    let mut out = vec![0u32; W * H];

    for _ in 0..N_WARMUP { render(&mut out); }

    let mut samples = [0.0; N_RUNS];
    for sample in &mut samples {
        let t0 = Instant::now();
        render(&mut out);
        *sample = t0.elapsed().as_secs_f64() * 1000.0;
    }

    println!(
        "median_us={} checksum={} samples={} stages={} runs={}",
        median_us(&mut samples),
        checksum_u32(&out),
        W * H,
        MAX_ITER,
        N_RUNS
    );
}
